File: sdk-typescript/src/context.ts
// Copyright (c) 2026 Super Durable, Inc.
//
// Licensed under the Super Durable Source License 1.0.
// You may not use this file except in compliance with the License.
// See the LICENSE file in the repository root.
//
// SPDX-License-Identifier: LicenseRef-Super-Durable-1.0

import type { Attribute, AttributeMap } from "./persistence.js";
import type { Channel, ChannelMap, ChannelMessage } from "./wait.js";
import type { Codec } from "./codec.js";
import type { Stream } from "./stream.js";

/**
 * Exposes execution metadata and decision-local persistence operations.
 * Dex supplies a Context to each Step or RPC handler; do not retain it afterward.
 */
export interface Context {
  /** Stable application Flow ID shared across runs. */
  readonly flowId: string;
  /** Current server-assigned run ID. */
  readonly runId: string;
  /** UTC timestamp at which the current Flow run started. */
  readonly flowStartedAt: Date;
  /** Current Step type and execution number encoded by Dex. */
  readonly stepExecutionId: string;
  /** Predecessor Step execution ID, or an empty string at Flow start. */
  readonly fromStepExecutionId: string;
  /** UTC timestamp of the first handler attempt. */
  readonly firstAttemptAt: Date;
  /** One-based handler retry attempt number. */
  readonly attempt: number;
  /**
   * Signal aborted when Dex cancels the active Worker call.
   *
   * Pass it to abort-aware APIs or inspect it at natural CPU-work boundaries. JavaScript
   * cannot forcibly stop synchronous code, and cancellation cannot make external effects atomic.
   */
  readonly cancellationSignal: AbortSignal;
  /**
   * Reports whether a previous regular-activity attempt recorded heartbeat details.
   * @returns Whether a last heartbeat Value is present.
   */
  hasLastHeartbeatValue(): boolean;
  /**
   * Decodes heartbeat details recorded by the previous regular-activity attempt.
   *
   * The codec must match the one passed to {@link AsyncContext.recordHeartbeat}. When omitted,
   * values use JSON encoding. Call {@link Context.hasLastHeartbeatValue} to distinguish an absent
   * Value from a present JSON null value, because both decode to `undefined`.
   *
   * @typeParam T - Expected heartbeat value type.
   * @param codec - Codec used to decode the Value; omitted for JSON.
   * @returns The decoded value, or `undefined` when no Value is present.
   */
  getLastHeartbeatValue<T = unknown>(codec?: Codec<T>): T | undefined;
  /**
   * Reports whether a Timer made the current Wait ready.
   * @param index - Optional zero-based Timer index; checks any Timer when omitted.
   * @returns Whether the selected Timer fired.
   */
  hasTimerFired(index?: number): boolean;
  /**
   * Reports whether `waitFor` failed before the current `execute` call.
   * @returns Whether failure policy proceeded to execution.
   */
  waitForMethodFailed(): boolean;
  /**
   * Stores process-local data for this Step execution; it is not durable.
   * @typeParam T - Local value type.
   * @param key - Non-empty execution-scoped key.
   * @param value - Value retained in worker memory.
   * @param codec - Codec used to serialize the local value when required.
   */
  setStepExecutionLocal<T>(key: string, value: T, codec: Codec<T>): void;
  /**
   * Reads process-local data for this Step execution.
   * @typeParam T - Expected local value type.
   * @param key - Key used when storing the value.
   * @param codec - Codec used to decode the value.
   * @returns The value, or `undefined` after absence or worker restart.
   */
  getStepExecutionLocal<T>(key: string, codec: Codec<T>): T | undefined;
  /**
   * Stages an application event in the current handler result.
   * @typeParam T - Event payload type.
   * @param name - Non-empty diagnostic event name.
   * @param value - Event payload.
   * @param codec - Payload codec.
   */
  recordEvent<T>(name: string, value: T, codec: Codec<T>): void;
  /**
   * Reads a typed Attribute from decision state.
   * @typeParam T - Attribute value type.
   * @param attribute - Registered singleton or map definition.
   * @param instance - Required AttributeMap instance; omitted for a singleton.
   * @returns The decoded current value.
   */
  getAttribute<T>(attribute: Attribute<T> | AttributeMap<T>, instance?: string): T;
  /**
   * Stages an Attribute write in the current decision.
   * @typeParam T - Attribute value type.
   * @param attribute - Registered singleton or map definition.
   * @param value - Typed value to persist.
   * @param instance - Required AttributeMap instance; omitted for a singleton.
   */
  setAttribute<T>(attribute: Attribute<T> | AttributeMap<T>, value: T, instance?: string): void;
  /**
   * Stages deletion of an Attribute value.
   * @param attribute - Registered singleton or map definition.
   * @param instance - Required AttributeMap instance; omitted for a singleton.
  */
  deleteAttribute(attribute: Attribute<unknown> | AttributeMap<unknown>, instance?: string): void;
  /**
   * Returns effective decoded AttributeMap keys in ascending order.
   * @param attribute - Registered AttributeMap definition.
   * @returns Keys visible after decision-local writes and deletions.
   */
  attributeMapKeys(attribute: AttributeMap<unknown>): readonly string[];
  /**
   * Stages one typed Channel publication.
   * @typeParam T - Channel element type.
   * @param channel - Registered singleton or map definition.
   * @param value - Value to append.
   * @param instance - Required ChannelMap instance; omitted for a singleton.
   */
  publish<T>(channel: Channel<T> | ChannelMap<T>, value: T, instance?: string): void;
  /**
   * Stages deletion of one pending Channel message from an RPC handler.
   * @param channel - Registered singleton or map definition.
   * @param messageId - Non-empty server-assigned message ID.
   * @param instance - Required ChannelMap instance; omitted for a singleton.
   */
  deleteChannelMessage(
    channel: Channel<unknown> | ChannelMap<unknown>,
    messageId: string,
    instance?: string,
  ): void;
  /**
   * Returns a Channel's current queued value count.
   * @param channel - Registered singleton or map definition.
   * @param instance - Required ChannelMap instance; omitted for a singleton.
   * @returns Non-negative queued value count.
   */
  channelSize(channel: Channel<unknown> | ChannelMap<unknown>, instance?: string): number;
  /**
   * Returns loaded pending Channel messages in FIFO order.
   * @typeParam T - Channel element type.
   * @param channel - Registered singleton or map definition selected by the RPC.
   * @param instance - Required ChannelMap instance; omitted for a singleton.
   * @returns Immutable pending message IDs and decoded values.
   */
  pendingChannelMessages<T>(
    channel: Channel<T> | ChannelMap<T>,
    instance?: string,
  ): readonly ChannelMessage<T>[];
  /**
   * Returns values selected by the satisfied Channel condition.
   * @typeParam T - Channel element type.
   * @param channel - Registered singleton or map definition.
   * @param instance - Required ChannelMap instance; omitted for a singleton.
   * @returns Ordered values for this Step execution.
   */
  channelResults<T>(channel: Channel<T> | ChannelMap<T>, instance?: string): readonly T[];
  /**
   * Returns decoded non-empty ChannelMap keys in ascending order during an RPC.
   * @param channel - Registered ChannelMap definition.
   * @returns Keys including publications buffered by the current RPC.
   */
  channelMapKeys(channel: ChannelMap<unknown>): readonly string[];
  /**
   * Appends one immediate best-effort Stream message.
   * @typeParam T - Stream message type.
   * @param stream - Stream registered by the current Flow type.
   * @param value - Typed message to append.
   */
  writeStream<T>(stream: Stream<T>, value: T): void;
}

/**
 * Adds asynchronous progress reporting to a Step invocation Context.
 *
 * Annotate an async `waitFor` or `execute` handler with AsyncContext when it records heartbeats.
 * RPC and Flow timeout handlers receive Context and cannot record Step heartbeats.
 */
export interface AsyncContext extends Context {
  /**
   * Records heartbeat progress for the current Step method attempt.
   *
   * Passing `undefined` explicitly clears persisted heartbeat details. Every other value produces
   * a present Value; omitted codecs use JSON. The Promise resolves when grpc-js accepts the frame,
   * not when Temporal persists it. Local activities ignore heartbeat frames.
   *
   * @typeParam T - Heartbeat value type.
   * @param value - Value to persist, or `undefined` to clear existing details.
   * @param codec - Codec used to encode the Value; omitted for JSON.
   * @returns A Promise resolved after the Worker accepts the output frame locally.
   */
  recordHeartbeat<T>(value: T | undefined, codec?: Codec<T>): Promise<void>;
}

File: sdk-java/src/main/java/io/superdurable/dex/Context.java
/*
 * Copyright (c) 2026 Super Durable, Inc.
 *
 * Licensed under the Super Durable Source License 1.0.
 * You may not use this file except in compliance with the License.
 * See the LICENSE file in the repository root.
 *
 * SPDX-License-Identifier: LicenseRef-Super-Durable-1.0
 */

package io.superdurable.dex;

import java.time.Instant;
import java.util.List;

/**
 * Exposes invocation metadata and durable state operations to Step and RPC code.
 *
 * <p>A context belongs to one worker invocation and must not be retained or used from another
 * thread after the method returns. Attribute and Channel helpers are also available directly on
 * their typed definitions; the context methods are useful for framework-style code. Values passed
 * with a {@link Class} must use concrete, non-parameterized types. Wrap a parameterized value in a
 * concrete holder class, or use an array such as {@code String[]} instead of {@code List<String>}.
 *
 * <pre>{@code
 * public StepDecision execute(Context context, Order input) {
 *     int attempt = context.getAttempt();
 *     status.set(context, "processed-on-attempt-" + attempt);
 *     context.recordEvent("order-processed", input, Order.class);
 *     return StepDecision.gracefulComplete();
 * }
 * }</pre>
 */
public interface Context {
    /**
     * Returns the Flow ID for the current invocation.
     *
     * @return the Flow ID
     */
    String getFlowId();

    /**
     * Returns the current Flow run ID.
     *
     * @return the run ID
     */
    String getRunId();

    /**
     * Returns when the current Flow execution started.
     *
     * @return the Flow start timestamp
     */
    Instant getFlowStartedAt();

    /**
     * Returns the server-assigned identifier of the current Step execution.
     *
     * @return the Step execution ID
     */
    String getStepExecutionId();

    /**
     * Returns the Step execution that scheduled the current Step.
     *
     * @return the originating Step execution ID, or an empty value when none exists
     */
    String getFromStepExecutionId();

    /**
     * Returns when the first attempt of the current method began.
     *
     * @return the first-attempt timestamp
     */
    Instant getFirstAttemptAt();

    /**
     * Returns the one-based attempt number of the current method invocation.
     *
     * @return the current attempt number
     */
    int getAttempt();

    /**
     * Reports whether any timer condition in the current wait has fired.
     *
     * @return {@code true} when at least one timer fired
     */
    boolean hasTimerFired();

    /**
     * Reports whether the timer at a specific condition index fired.
     *
     * @param index the zero-based timer condition index
     * @return {@code true} when that timer fired
     */
    boolean hasTimerFired(int index);

    /**
     * Reports whether execute is running after a failed wait-for method.
     *
     * <p>This can be {@code true} only when the Step uses
     * {@link WaitForFailurePolicy#PROCEED}.
     *
     * @return {@code true} when wait-for retries were exhausted before execute
     */
    boolean waitForMethodFailed();

    /**
     * Stores a value scoped to the current Step execution.
     *
     * <p>Step-execution locals allow {@link Step#waitFor} to pass durable data to
     * {@link Step#execute} for the same execution.
     *
     * @param key the local value key
     * @param value the value to serialize
     * @param valueType the concrete Java value class
     * @param <T> the local value type
     */
    <T> void setStepExecutionLocal(String key, T value, Class<T> valueType);

    /**
     * Reads a value scoped to the current Step execution.
     *
     * @param key the local value key
     * @param valueType the concrete Java class used to decode the value
     * @param <T> the local value type
     * @return the decoded value, or {@code null} when the key is absent
     */
    <T> T getStepExecutionLocal(String key, Class<T> valueType);

    /**
     * Appends a named user event to the Flow execution history.
     *
     * @param name the event name
     * @param value the event payload
     * @param valueType the concrete Java payload class
     * @param <T> the event payload type
     */
    <T> void recordEvent(String name, T value, Class<T> valueType);

    /**
     * Reads an Attribute.
     *
     * @param attribute the registered typed Attribute
     * @param <T> the Attribute value type
     * @return the current value, or {@code null} when absent
     */
    <T> T getAttribute(Attribute<T> attribute);

    /**
     * Reads one Attribute-map instance.
     *
     * @param attribute the registered typed Attribute map
     * @param instance the map instance
     * @param <T> the Attribute value type
     * @return the current value, or {@code null} when absent
     */
    <T> T getAttribute(AttributeMap<T> attribute, String instance);

    /**
     * Writes an Attribute.
     *
     * @param attribute the registered typed Attribute
     * @param value the value to persist
     * @param <T> the Attribute value type
     */
    <T> void setAttribute(Attribute<T> attribute, T value);

    /**
     * Writes one Attribute-map instance.
     *
     * @param attribute the registered typed Attribute map
     * @param instance the map instance
     * @param value the value to persist
     * @param <T> the Attribute value type
     */
    <T> void setAttribute(AttributeMap<T> attribute, String instance, T value);

    /**
     * Deletes an Attribute.
     *
     * @param attribute the registered Attribute
     */
    void deleteAttribute(Attribute<?> attribute);

    /**
     * Deletes one Attribute-map instance.
     *
     * @param attribute the registered Attribute map
     * @param instance the map instance
     */
    void deleteAttribute(AttributeMap<?> attribute, String instance);

    /**
     * Publishes one message to a Channel.
     *
     * @param channel the registered typed Channel
     * @param value the message value
     * @param <T> the Channel message type
     */
    <T> void publish(Channel<T> channel, T value);

    /**
     * Publishes one message to a Channel-map instance.
     *
     * @param channel the registered typed Channel map
     * @param instance the map instance
     * @param value the message value
     * @param <T> the Channel message type
     */
    <T> void publish(ChannelMap<T> channel, String instance, T value);

    /**
     * Returns the visible message count for a Channel.
     *
     * @param channel the registered Channel
     * @return the visible message count
     */
    int channelSize(Channel<?> channel);

    /**
     * Returns the visible message count for one Channel-map instance.
     *
     * @param channel the registered Channel map
     * @param instance the map instance
     * @return the visible message count
     */
    int channelSize(ChannelMap<?> channel, String instance);

    /**
     * Returns condition results for a Channel.
     *
     * @param channel the registered typed Channel
     * @param <T> the Channel message type
     * @return the typed messages consumed by the satisfied condition
     */
    <T> List<T> channelResults(Channel<T> channel);

    /**
     * Returns condition results for one Channel-map instance.
     *
     * @param channel the registered typed Channel map
     * @param instance the map instance used by the condition
     * @param <T> the Channel message type
     * @return the typed messages consumed by the satisfied condition
     */
    <T> List<T> channelResults(ChannelMap<T> channel, String instance);
}
